Change `NHWC`/`NCHW` to `NWC`/`NCW` for conv1d

yongtang opened the pull request:
While working on #13105 I noticed that in the current code base `conv1d` uses `NHWC`/`NCHW` which should really be `NWC`/`NCW`.

This fix addresses this issue and keep `NHWC`/`NCHW` compatible internally so that users will not be impacted.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Change `NHWC`/`NCHW` to `NWC`/`NCW` for conv1d

While working on 13105 I noticed that in the current code
base `conv1d` uses `NHWC`/`NCHW` which should really
be `NWC`/`NCW`.

This fix addresses this issue and keep `NHWC`/`NCHW`
compatible internally so that users will not be impacted.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

tensorflow-jenkins commented:
Can one of the admins verify this patch?

jhseu commented:
Not sure we'll be able to change this since it's not backwards-compatible.

martinwicke commented on tensorflow/python/ops/nn_ops.py:
```diff
@@ -2202,9 +2202,9 @@ def conv1d(value, filters, stride, padding,
-    data_format: An optional `string` from `"NHWC", "NCHW"`.  Defaults
-      to `"NHWC"`, the data is stored in the order of
-      [batch, in_width, in_channels].  The `"NCHW"` format stores
+    data_format: An optional `string` from `"NWC", "NCW"`.  Defaults
```
We cannot change the signature this way, as @jhseu said, this is a backwards incompatible change. Instead, you could accept both NHWC and NWC. You can even only document NWC to push users away from NHWC. Optionally, you can use @deprecated_arg_values to mark NWHC explicitly deprecated (will trigger a warning if used).

yongtang replied:
Thanks @martinwicke @jhseu for the review. The current PR actually accept both `NWC` and `NHWC`, (see Ln 2219):
```
    if data_format is None or data_format == "NHWC" or data_format == "NWC":
      data_format = "NHWC"
      spatial_start_dim = 1
      strides = [1, 1, stride, 1]
    elif data_format == "NCHW" or data_format == "NCW":
      data_format = "NCHW"
      spatial_start_dim = 2
      strides = [1, 1, 1, stride]
    else:
```

I think `@deprecated_arg_values` is good. Will update the PR to add it.

yongtang committed:
Add @deprecated_arg_values to conv1d for `NHWC/NCHW`

This commit adds @deprecated_arg_values to conv1d for `NHWC/NCHW`
(use `NWC/NCW` instead)

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @martinwicke for the tip of `@deprecated_arg_values`. The PR has been updated with `@deprecated_arg_values` added. Please take a look.

martinwicke commented:
Jenkins, test this please.

asimshankar commented:
(From API review)

Is using "width" for 1D somehow better than "height"? (Or the other way around).
Sounds like `W` or `H` are also inappropriate for 1D values.

I'd be tempted to leave things as-is for now?

yongtang commented:
Thanks @asimshankar for the review. The NCW/NWC was mainly to following other ops in nn like `convolution`:
https://www.tensorflow.org/api_docs/python/tf/nn/convolution


data_format: A string or None. Specifies whether the channel dimension of the input and output is the last dimension (default, or if data_format does not start with "NC"), or the second dimension (if data_format starts with "NC"). For N=1, the valid values are "NWC" (default) and "NCW". For N=2, the valid values are "NHWC" (default) and "NCHW". For N=3, the valid values are "NDHWC" (default) and "NCDHW".


I also noticed that a similar op `conv1d_transpose` (as opposed to `conv1d` in this PR) is using `NCW/NWC` (in Ln 2328 of the same file). That was why I opened this PR.

Not sure what is the best way to handle the inconsistencies but happy to do one way or another.

martinwicke commented:
Ok, checked with @asimshankar, this is fine.
